src/tir/transforms/make_packed_api.cc

```diff
@@ -41,6 +41,56 @@
 namespace tvm {
 namespace tir {
 
+class ReturnRewriter : public StmtMutator {
+ public:
+  explicit ReturnRewriter(Var ret_var, Var ret_tcode) : ret_var_(ret_var), ret_tcode_(ret_tcode) {}
+
+  Stmt VisitStmt_(const EvaluateNode* node) override {
+    Stmt ret = StmtMutator::VisitStmt_(node);
+    const EvaluateNode* eval = ret.as<EvaluateNode>();
+    CHECK(eval);
```

Comment:
Please change these to ICHECK because they are internal errors. Also, please add an error message. Something along the lines of "Expected EvaluateNode, but got ...."

Reply:
in this particular case because it is a code invariant(StmtMutator returns the same type). So changing to ICHECK without error message should be fine

Reply:
We always say things are invariant, yet I've hit many of them. It doesn't take much time to add a helpful error message

Reply:
I would agree for cases where the invariant was less clear. In this particular case the impl of StmtMutator was quite direct and invariant is clear, so adding error message won't buy us too much in this particular case. Additionally, having to have a long error message for all cases would increase the binary size.

I think we should certainly deliberate error messages that we do hit. One way to do so is to re-actively add them when we hit the problem. This case is certainly fine

Reply:
As someone who has hit many of these errors, an error message goes a long way towards helping me debug these problems. Even if the error message is as simple as "Internal invariant violated: an EvaluateNode was expected, but we got a ___". The message "CHECK(eval)" tells me nothing.

I don't think the argument for binary size maxes much of a difference here. We already include a string for the CHECK statement, this just makes it marginally larger.

Reply:
I agree with Tristan, I think we should consider bare `CHECK` statements as a code-smell as any new users trying to understand why an invariant doesn't hold have zero context, given the fact that some are invariants and some are just lazy error handling we should better differentiate which is which and not leave it up to the user.

Reply:
Thanks @tkonolige. 

In this particular case, the error message point to this line, and by looking around the context, one line above, we know that the error is caused by the type, so the context gives enough info. Notably, from the impl of StmtMutator this is an internal error that won't trigger and the check here is mainly a pre-caution.

I agree that in some other cases the context is certainly not enough for debug, and having a better error message would help. While adding error messages is certainly fine in this case, I do think not putting constraint on development in this particular case is helpful and @ZihengJiang can make the decision of what level of error message to be put in this internal check.

Reply:
Given that informations like "internal invariant violated: an EvaluateNode was expected, but we got a ___".  is quite boilerplate, maybe `ICHECK` should be improved instead of write error message in every place.

Reply:
Yeah, we certainly could add another `ICHECK` macro for that.

To be honest though, I don't really understand this invariant. Does the base class (StmtMutator) always return the same type of node you pass into it?

Reply:
yes, the base class always return the same type in this case

Reply:
It is important to take the context into consideration. For the checks that would leads to a user facing error, e.g. the API expects inputs of two elements but only get one, a more comprehensive error message should be used -- and we should use CHECK instead of ICHECK in the user facing API part.

In cases like this one, this is an invariant, and we could give the developer the freedom to choose the appropriate error message, since the ease of development is also another factor we want to balance.